blsftcosi/protocol/sub_protocol.go

```diff
@@ -0,0 +1,431 @@
+				if err := bls.Verify(p.suite, public, p.Msg, reply.Signature); err == nil {
+					responses[reply.ID] = &reply.Response
+					done++
+				}
+			} else {
+				log.Warnf("Duplicate message from %v", reply.ServerIdentity.ID)
+			}
+		case reply := <-p.ChannelRefusal:
+			_, ok := responses[reply.ID]
+			serviceName := onet.ServiceFactory.Name(p.Token().ServiceID)
+
+			if !ok {
+				log.Warnf("Got a message from an unknown node %v", reply.ServerIdentity.ID)
+			} else if err := bls.Verify(p.suite, reply.ServerIdentity.ServicePublic(serviceName), reply.Nonce, reply.Signature); err == nil {
+				// The child gives an empty signature as a mark of refusal
+				responses[reply.ID] = &Response{}
+				done++
+			} else {
+				log.Warnf("Tentative to send a unsigned refusal from %v", reply.ServerIdentity.ID)
+			}
+		case <-timeout:
+			// Use whatever we received until then to try to finish
+			// the protocol
+			done = len(p.Children())
+		}
+	}
+
+	r, err := makeAggregateResponse(p.suite, p.Publics(), responses)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	log.Lvlf3("Subleader %v sent its reply with mask %b", p.ServerIdentity(), r.Mask)
+	return p.SendToParent(r)
+}
+
+// dispatchLeaf prepares the signature and send it to the subleader
+func (p *SubBlsFtCosi) dispatchLeaf() error {
+	a := p.waitAnnouncement(p.Root().Children[0])
+	if a == nil {
+		return nil
+	}
+
+	res := make(chan bool)
+	go p.makeVerification(res)
+
+	// give a chance to avoid sending the response if a stop
+	// has been requested
+	select {
+	case <-p.closeChan:
+		return nil
+	case ok := <-res:
+		var r interface{}
+		var err error
+		if ok {
+			log.Lvlf3("Leaf %v signed", p.ServerIdentity())
+			r, err = p.makeResponse()
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Lvlf3("Leaf %v refused to sign", p.ServerIdentity())
+			r, err = p.makeRefusal()
+			if err != nil {
+				return err
+			}
+		}
+
+		return p.SendToParent(r)
+	}
+}
+
+// Sign the message and pack it with the mask as a response
+func (p *SubBlsFtCosi) makeResponse() (*Response, error) {
+	mask, err := cosi.NewMask(p.suite, p.Publics(), p.Public())
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+
+	sig, err := bls.Sign(p.suite, p.Private(), p.Msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Response{
+		Mask:      mask.Mask(),
+		Signature: sig,
+	}, nil
+}
+
+// makeRefusal will sign a random nonce so that we can check
+// that the refusal is not forged
+func (p *SubBlsFtCosi) makeRefusal() (*Refusal, error) {
+	nonce := make([]byte, 8)
+	_, err := rand.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	sig, err := bls.Sign(p.suite, p.Private(), nonce)
+
+	return &Refusal{Signature: sig, Nonce: nonce}, err
+}
+
+// makeAggregateResponse takes all the responses from the children and the subleader to
+// aggregate the signature and the mask
+func makeAggregateResponse(suite pairing.Suite, publics []kyber.Point, responses ResponseMap) (*Response, error) {
```

Comment:
is it possible to simplify this method using what is in https://github.com/dedis/kyber/pull/338/files ?

Reply:
I don't think it's worth it. You still need to aggregate the mask so the loop can't be avoided and thus you save maybe one line.